Add BidDelegate integration example with Facebook Audience Network

lovell opened the pull request:
Hello, this PR adds code to help publishers integrate Pubfood.js with Facebook Audience Network by providing a new `AudienceNetworkBidDelegate`.

I've taken the liberty of using ES6 for the source as well as including the transpiled, minified ES5 version too. There are simple-to-follow instructions in the README to re-build this should changes be required. I suspect some publishers might want to import the ES6 version when generating their own aggregated JavaScript bundle. This PR provides for both options.

It includes unit tests with 100% branch coverage.

I am not a Facebook employee, but this work was commissioned and paid for by Facebook.

elljoh commented:
@lovell Thanks for the contribution and diligence the pubfood team are reviewing and I'll mention the PR in the community channels for review/feedback also.

Cheers

elljoh commented on examples/provider/bid/audience-network/src/audience-network-bid-delegate.js:
```diff
@@ -0,0 +1,134 @@
+const validateSlotSizes = slot => {
+  slot.sizes = slot.sizes.map(flattenSize);
+  return slot.sizes.every( size =>
+    ['native', 'fullwidth', '300x250', '320x50'].includes(size) );
```
For the so-called multi-size slot such as in the example here, [examples/provider/bid/ix/ix-ex1.html#L14](https://github.com/pubfood/pubfood/blob/master/examples/provider/bid/ix/ix-ex1.html#L14), should we interpret this such that FB will not bid?

Mainly curiosity.

lovell replied:
Yes, based on my understanding, the sizes that Audience Network will provided valid bids for is limited to those included here.

elljoh replied:
Ok. I do wonder if using `Array.prototype.every()` here makes things a little restrictive i.e. it's not uncommon for publishers to define adunits such that they will render, for example, either a 300x250 or 300x600.

A bid for only the 300x250 dimension would be just fine.

wheresbarney replied:
Agree - we saw the same issue as soon as people started using the Prebid.js equivalent, and I think it was switched to a filter instead. We should bid on the first size we recognise.

elljoh commented on examples/provider/bid/audience-network/README.md:
````diff
@@ -0,0 +1,46 @@
+
+## Usage
+
+```html
````
This snippet is a good start; however, it is desirable that there is an example page a consumer of the Pubfood repo can load either via local web server and/or filesystem.

In principle, what we see in the _audience-network-bid-delegate.js_ module "should" :tm: work, but generally good to prove that out with a simple integration example page even if Audience Network does not bid.

* On this test Audience Network bids, must a consumer of the repo have their own Audience Network account and placementId or does FB provide a test `{accountId}_{placementId}` for peoples to use?

Like in other header bidding framework adapter modules, it appears that a consumer not only needs an Audience Network account and placementId, there needs to be some additional flag set on the Audience Network system for the Account/Placement?

* `https://an.facebook.com/v2/placementbid.json?sdk=5.5.web&placementids[]=1798405500476623_1798416553808851&adformats[]=300x250`
* `errors:["Application not authorised for header bidding"]`

Could I ask:
1. Create an integration test page
* Perhaps you can base the integration on something from here [Audience Network examples](https://github.com/fbsamples/audience-network/tree/master/samples/mobile_web)

2. Look into a public sandbox/demo `placementId` for use in the example
* The "5...'s" placementId is fine for the commit, but could you enable the placementId above for header bidding if FB do not have a sandbox/demo identifier to use

The primary object with the example integration page is to ensure the respective adapter implementation example bids will indeed flow through the framework and to the AuctionProvider delegate.

lovell replied:
I believe that, currently, header bidding via Audience Network requires a manual approval step, hence the "not authorised for header bidding" message you are experiencing. Let me get someone with the relevant rights to approve your test account,

wheresbarney replied:
@elljoh - I've whitelisted your test app for AN Header Bidding - can you retest please?

elljoh replied:
@wheresbarney thanks.

I do not see the header bidding auth error. This is what I see now:

```
{
   "request_id": "6261999187131519857",
   "errors": [
      "1798405500476623_1798416553808851 : "
   ],
   "bids": {
      
   }
}
```

My test app setup is pretty naive. From what you see is there any config or setup I am missing in order to receive test bids?

I do see this message in the AN UI console:

> Placement ID:	1798405500476623_1798416553808851
> Status:	Enabled for testing and received ad request recently

wheresbarney replied:
@elljoh - can you check you were signed in to facebook.com in the same browser session, and that you were using a mobile user agent, please?

elljoh replied:
@wheresbarney great, thanks. mobile UA was the issue it seems.

```
{
   "request_id": "2529434687361507514",
   "errors": [
      
   ],
   "bids": {
      "1798405500476623_1798416553808851": [
         {
            "placement_id": "1798405500476623_1798416553808851",
            "bid_id": "23112430200564801",
            "bid_price_cents": 369.6,
            "bid_price_currency": "usd",
            "bid_price_model": "cpm"
         }
      ]
   }
}
```

elljoh commented on examples/provider/bid/audience-network/src/audience-network-bid-delegate.js:
```diff
@@ -0,0 +1,134 @@
+ * @returns {Boolean}
+ */
+const validateSlotSizes = slot => {
+  slot.sizes = slot.sizes.map(flattenSize);
```
The `slot` here is a reference i.e. slot sizes configuration passed to `addSlot()`, as in the README example `sizes: [[300, 250]]`, will be redefined.

At this time, Pubfood does not pass cloned Slot objects to delegates. This function, setting of `adformats` and the `pushBid` argument for sizes will need to reflect this.

lovell commented:
@elljoh Great to see you got your sample placement working. I plan to update this PR with the two suggested changes for slot size filtering plus some additional "quick start" help, most likely early next week.

lovell commented:
@elljoh A few updates:

* incoming `slots` passed to `init()` / `refresh()` are now considered immutable
* any valid slot size will be bid on, with invalid slot sizes filtered before bidding
* added a "quick start" guide to creating a new `placementId`
* removed `lodash` dependency for ~20% smaller minified version (3.4KB when gzip-compressed)

elljoh commented on examples/provider/bid/audience-network/README.md:
```diff
@@ -0,0 +1,51 @@
+2. Add the Audience Network product to it
+3. Create a new Placement to generate your `placementId`
+4. To test, ensure the `User-Agent` request header represents a mobile device
+
```
Will the FB UI indicate that a `placementId` needs to be enabled for header bidding in the future? If not, perhaps a comment on how to action the `"Application not authorised for header bidding"` response message.

Good addition however.

lovell replied:
@wheresbarney should be able to confirm/deny this.

wheresbarney replied:
We only bid on approved publishers. The readme can mention a publisher should speak to their account manager to start the approval process.

elljoh commented on examples/provider/bid/audience-network/src/audience-network-bid-delegate.js:
```diff
@@ -0,0 +1,138 @@
+  const sizes = Array.isArray(bid.sizes) && bid.sizes
+    .map(flattenSize)
+    .filter(isValidSize);
+  return Object.assign({}, bid, { sizes });
```
`Object.assign` for IE is primarily an Edge browser capability IE10/9 will have an issue with this.

_See also_ comment for function `validateSlot`

elljoh commented on examples/provider/bid/audience-network/src/audience-network-bid-delegate.js:
```diff
@@ -0,0 +1,138 @@
+  typeof slot.elementId === 'string' &&
+  Array.isArray(slot.sizes) &&
+  slot.sizes.length > 0 &&
+  typeof slot.getParam === 'function' &&
```
a `Slot` inherits `getParam()` from `PubfoodObject` (_http://pubfood.org/api-reference#pubfood-model-Slot-getParam_)

`slot.getParam` here will be undefined.

`Object.assign` below will copy only enumerable _**owned**_ properties: being inherited, `getParam` will not be copied with `Object.assign`.

`Object.create` would do something more like you might want i.e. this will copy the prototype.

lovell replied:
Ah, I completely forgot this object will have inherited properties, thanks for spotting this. I'll update to flatten the `sizes` property when accessed rather than attempt to clone the `slot`.

wheresbarney commented:
Hi @lovell , @elljoh - where are we with this one? Any issues I can help with? Thanks!

lovell commented:
I think all the comments have now been addressed, e.g. there's no longer any mutation/cloning of `Slot` instances, plus I've rebased against the latest `master` as of yesterday.

elljoh commented:
Yes. I think were good to go. Will do one final pass through before a merge today.

elljoh commented:
LGTM, verified seeing expected slot custom parameter targeting in a publisher adserver (_DFP_) request e.g.
`hb_bidder=fan&hb_pb=1.40&fb_placementid=1798405500476623_1798416553808851&fb_bidid=4234990049101360193&fb_format=300x250`
